# Q1

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_excel("ENB2012_data.xlsx")
df

X1     X2     X3      X4   X5  X6   X7  X8     Y1     Y2
0    0.98  514.5  294.0  110.25  7.0   2  0.0   0  15.55  21.33
1    0.98  514.5  294.0  110.25  7.0   3  0.0   0  15.55  21.33
2    0.98  514.5  294.0  110.25  7.0   4  0.0   0  15.55  21.33
3    0.98  514.5  294.0  110.25  7.0   5  0.0   0  15.55  21.33
4    0.90  563.5  318.5  122.50  7.0   2  0.0   0  20.84  28.28
..    ...    ...    ...     ...  ...  ..  ...  ..    ...    ...
763  0.64  784.0  343.0  220.50  3.5   5  0.4   5  17.88  21.40
764  0.62  808.5  367.5  220.50  3.5   2  0.4   5  16.54  16.88
765  0.62  808.5  367.5  220.50  3.5   3  0.4   5  16.44  17.11
766  0.62  808.5  367.5  220.50  3.5   4  0.4   5  16.48  16.61
767  0.62  808.5  367.5  220.50  3.5   5  0.4   5  16.64  16.03

[768 rows x 10 columns]

In [ ]:
X = df.drop(["Y1","Y2"],axis=1)
Y = df[["Y1","Y2"]]
print(X.shape,Y.shape)

(768, 8) (768, 2)


In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2)
print(X_train.shape,Y_train.shape)

(614, 8) (614, 2)


In [ ]:
from sklearn.linear_model import LinearRegression
LM = LinearRegression()
LM.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
LM.coef_

array([[-6.57419290e+01, -6.29200285e-02,  3.43264217e-02,
        -4.86232251e-02,  4.25874094e+00,  5.51909197e-02,
         1.98587297e+01,  2.08127917e-01],
       [-7.34661301e+01, -6.84581818e-02,  2.10785395e-02,
        -4.47683607e-02,  4.33995310e+00,  2.15869175e-01,
         1.47305753e+01,  3.39072671e-02]])

In [ ]:
LM.intercept_

array([ 84.71487893, 100.8531584 ])

In [ ]:
Y_pred = LM.predict(X_test)
Y_pred = pd.DataFrame(Y_pred,columns=["Y1","Y2"])
Y_pred

Y1         Y2
0    34.708348  36.265831
1     8.480815  12.264858
2    25.100277  28.629995
3    33.959294  36.028789
4    12.714193  15.458749
..         ...        ...
149  33.751166  35.994882
150  34.445029  36.016054
151  32.197385  33.772565
152  12.259892  15.233943
153  26.447894  29.387299

[154 rows x 2 columns]

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
mse = mean_squared_error(Y_pred,Y_test)
r2 = r2_score(Y_test,Y_pred)
print(mse,r2)

9.105443276750915 0.896631271184553


In [ ]:
for test_split in [0.4,0.3,0.2,0.1]:
  print("Test Split = "+str(test_split))
  X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=test_split)
  LM.fit(X_train,Y_train)
  print(f"Model coefficients: {LM.coef_.tolist()}")
  print(f"Intercepts: {LM.intercept_.tolist()}")
  Y_pred = LM.predict(X_test)
  display(pd.DataFrame(Y_pred,columns=["Y1","Y2"]))
  mse,r2 = mean_squared_error(Y_pred,Y_test),r2_score(Y_test,Y_pred)
  print("MSE = "+str(mse),"R2 = "+str(r2))
  print()

Test Split = 0.4
Model coefficients: [[-65.72288355404596, 936113739497.8496, -936113739497.8776, -1872227478995.8794, 4.027735710144043, -0.08740249276161194, 20.086064521339722, 0.24751383066177368], [-64.24627944035953, 225562494797.23422, -225562494797.26837, -451124989594.62427, 4.385019540786743, 0.07323663681745529, 15.747675053891726, 0.0420500785112381]]
Intercepts: [86.98719565217391, 85.02742934782609]


Y1         Y2
0    10.001633  13.097813
1     9.884092  13.393538
2    31.515275  32.159247
3    32.487395  34.767297
4    32.822311  34.736110
..         ...        ...
303  34.920137  35.988950
304   6.828167  11.501318
305  18.687564  19.906152
306   9.328167  13.470068
307  37.417282  37.852943

[308 rows x 2 columns]

MSE = 8.797564040827249 R2 = 0.9041329528325918

Test Split = 0.3
Model coefficients: [[-67.9685200429918, -58739158572.69378, 58739158572.66376, 117478317145.20831, 4.246674835681915, -0.002099666220601648, 20.32013070897665, 0.17633625026792288], [-74.94520752890317, 279318172755.21423, -279318172755.2626, -558636345510.6139, 4.341697692871094, 0.16092634992673993, 14.887912506470457, 0.06799376755952835]]
Intercepts: [87.94869297020483, 103.01668296089385]


Y1         Y2
0    10.101758  13.904158
1    11.250555  14.421158
2    28.412287  31.238133
3    11.248455  14.582085
4    12.362588  15.746499
..         ...        ...
226  16.567009  18.013388
227  24.036928  28.299258
228  33.287997  34.898727
229  30.954387  31.725960
230  31.483395  31.929942

[231 rows x 2 columns]

MSE = 8.522043479021082 R2 = 0.9062149799241592

Test Split = 0.2
Model coefficients: [[-70.49754673028492, -0.06820727088557821, 0.03470425105482234, -0.05145576101499866, 4.203642788764064, 0.007621897820041175, 19.5922133122515, 0.24487589117611103], [-77.11869141803673, -0.07106346808147565, 0.01954328655374136, -0.045303377366628816, 4.434400718974638, 0.13748315149322538, 14.435467778599882, 0.06449764943952604]]
Intercepts: [92.70867452754946, 105.76840408121163]


Y1         Y2
0    15.033571  17.847563
1    35.130292  36.257056
2    29.955128  32.342218
3    11.605398  14.716048
4    40.427173  40.225262
..         ...        ...
149  17.406552  19.308004
150  31.946584  34.027238
151  33.856056  34.752589
152  30.848736  32.012364
153  18.317065  19.391316

[154 rows x 2 columns]

MSE = 11.1675187031461 R2 = 0.8799382744473503

Test Split = 0.1
Model coefficients: [[-63.254152780772756, -0.062084337800005523, 0.037044656702402784, -0.04956449724203526, 4.1065380615011415, -0.045086218336593764, 19.750872786598794, 0.21821075263080386], [-69.54314725715426, -0.06441702615008243, 0.021744567129276773, -0.04308079662955776, 4.3374570261980505, 0.09784637742920904, 14.225168590180528, 0.06790936596940829]]
Intercepts: [82.60391902153475, 95.00013183248059]


Y1         Y2
0   36.758983  37.760387
1   15.182499  17.446901
2   40.248208  39.834289
3   35.362857  36.830227
4   17.192357  18.805931
..        ...        ...
72   9.719645  13.527498
73  28.153443  31.521857
74  26.500046  29.755669
75  22.754695  25.906424
76  15.866139  17.721298

[77 rows x 2 columns]

MSE = 8.671413692205629 R2 = 0.90675085142135



Least MSE (8.522) obtained for Test Split = 0.3 & Highest R2 (0.9067) obtained for Test Split = 0.1

In [ ]:
from sklearn.linear_model import Lasso,Ridge

In [ ]:
# Ridge Regression Model
for alpha in [0.001,0.005,0.01,0.05,0.1,0.5]:
  print("Alpha = "+str(alpha))
  for test_split in [0.4,0.3,0.2,0.1]:
    X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = test_split)
    RM = Ridge(alpha=alpha)
    RM.fit(X_train,Y_train)
    Y_pred = RM.predict(X_test)
    mse = mean_squared_error(Y_pred,Y_test)
    r2 = r2_score(Y_test,Y_pred)
    print(f'Test split: {test_split}  MSE: {mse}  R2: {r2}')

Alpha = 0.001
Test split: 0.4  MSE: 8.652562410794474  R2: 0.9066987316137767
Test split: 0.3  MSE: 8.626648642894242  R2: 0.9069534293646502
Test split: 0.2  MSE: 9.250243709747352  R2: 0.8968988978434456
Test split: 0.1  MSE: 9.364011026519673  R2: 0.8834628243207825
Alpha = 0.005
Test split: 0.4  MSE: 9.435247912235594  R2: 0.901430378144529
Test split: 0.3  MSE: 10.162255886124239  R2: 0.8949196625998077
Test split: 0.2  MSE: 10.620985083014759  R2: 0.8910725281721887
Test split: 0.1  MSE: 8.112631393588297  R2: 0.9242674292107275
Alpha = 0.01
Test split: 0.4  MSE: 8.318503412491658  R2: 0.9038320765998695
Test split: 0.3  MSE: 8.641821566385229  R2: 0.9098720580316777
Test split: 0.2  MSE: 10.301839331111003  R2: 0.8841602556226655
Test split: 0.1  MSE: 10.429882181246633  R2: 0.8836805674243211
Alpha = 0.05
Test split: 0.4  MSE: 9.59150825417337  R2: 0.8975331411230854
Test split: 0.3  MSE: 9.589849416267857  R2: 0.9019497656748244
Test split: 0.2  MSE: 11.544299410201965  R2: 0.

Least MSE = 8.11 & Best R2 = 0.924 are obtained for same Alpha = 0.005, Test Split = 0.1

In [ ]:
# Lasso Regression Model
import warnings
warnings.filterwarnings("ignore") # Warnings are suppressed for better view of print statements
for alpha in [0.001,0.005,0.01,0.05,0.1,0.5]:
  print("Alpha = "+str(alpha))
  for test_split in [0.4,0.3,0.2,0.1]:
    X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = test_split)
    LsM = Lasso(alpha=alpha)
    LsM.fit(X_train,Y_train)
    Y_pred = LsM.predict(X_test)
    mse = mean_squared_error(Y_pred,Y_test)
    r2 = r2_score(Y_test,Y_pred)
    print(f'Test split: {test_split}  MSE: {mse}  R2: {r2}')

Alpha = 0.001
Test split: 0.4  MSE: 9.569998869637157  R2: 0.8926034816828632
Test split: 0.3  MSE: 9.82906561182016  R2: 0.8979214823755269
Test split: 0.2  MSE: 9.057231603285137  R2: 0.9146120450305139
Test split: 0.1  MSE: 5.533429873233947  R2: 0.9398763475276659
Alpha = 0.005
Test split: 0.4  MSE: 8.7384509265052  R2: 0.904543073441608
Test split: 0.3  MSE: 9.383758143326503  R2: 0.9021241585399096
Test split: 0.2  MSE: 8.576074520777624  R2: 0.9047894786640462
Test split: 0.1  MSE: 9.186042834436718  R2: 0.9105041525815323
Alpha = 0.01
Test split: 0.4  MSE: 9.669710438348012  R2: 0.8946474915958431
Test split: 0.3  MSE: 10.050864639021668  R2: 0.8847781478702391
Test split: 0.2  MSE: 9.395018442896617  R2: 0.9051608299146443
Test split: 0.1  MSE: 9.748252777226185  R2: 0.9084926264633166
Alpha = 0.05
Test split: 0.4  MSE: 10.387138921252753  R2: 0.8928877918081265
Test split: 0.3  MSE: 10.32728743062343  R2: 0.8895870101080955
Test split: 0.2  MSE: 8.464808428487082  R2: 0.90335

Least MSE = 5.53 & Best R2 = 0.94 are obtained for same Alpha = 0.001 & Test Split = 0.1

Lasso Regression Model has yielded least Mean Square Error and best value of R^2

# Q2

In [ ]:
df2 = pd.read_table("haberman.data",delimiter=",",names=["Age","Year","Nodes","Status"])
df2

Age  Year  Nodes  Status
0     30    64      1       1
1     30    62      3       1
2     30    65      0       1
3     31    59      2       1
4     31    65      4       1
..   ...   ...    ...     ...
301   75    62      1       1
302   76    67      0       1
303   77    65      3       1
304   78    65      1       2
305   83    58      2       2

[306 rows x 4 columns]

In [ ]:
X2 = df2.drop(["Status"],axis=1)
Y2 = df2[["Status"]]
print(X2.shape,Y2.shape)

(306, 3) (306, 1)


In [ ]:
X2_train,X2_test,Y2_train,Y2_test = train_test_split(X2,Y2,test_size = 0.2)
print(X2_train.shape,Y2_train.shape)

(244, 3) (244, 1)


In [ ]:
from sklearn.linear_model import LogisticRegression
LgM = LogisticRegression()
LgM.fit(X2_train,Y2_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
LgM.coef_

array([[0.02127065, 0.00310661, 0.07054207]])

In [ ]:
LgM.intercept_

array([-2.63742261])

In [ ]:
Y2_pred = LgM.predict(X2_test)
Y2_pred_df = pd.DataFrame(Y2_pred,columns=["Status"])
Y2_pred_df

Status
0        1
1        1
2        1
3        1
4        1
..     ...
57       1
58       1
59       1
60       1
61       1

[62 rows x 1 columns]

In [ ]:
# Indicator function
def indicator(a,b):
  ind=[1 if a[i]!=b[i] else 0 for i in range(len(a))]
  return np.array(ind)

In [ ]:
# Classification Error
ind = indicator(Y2_pred,np.array(Y2_test))
ce = (np.sum(ind))/len(Y2_pred)
ce

0.20967741935483872

In [ ]:
for test_split in [0.4,0.3,0.2,0.1]:
  print("Test Split = "+str(test_split))
  X2_train,X2_test,Y2_train,Y2_test = train_test_split(X2,Y2,test_size = 0.2)
  LgM.fit(X2_train,Y2_train)
  print(f"Model coefficients: {LgM.coef_.tolist()[0]}")
  print(f"Intercept = {LgM.intercept_.tolist()[0]}")
  Y2_pred = LgM.predict(X2_test)
  print(f"Predicted Statuses: {Y2_pred.tolist()}")
  ind = indicator(Y2_pred,np.array(Y2_test))
  ce = (np.sum(ind))/len(Y2_pred)
  print("Classification Error = "+str(ce))
  print()

Test Split = 0.4
Model coefficients: [0.024315382052933744, 0.016188012261961492, 0.10595270306103857]
Intercept = -3.852435261161604
Predicted Statuses: [1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 2, 1, 1]
Classification Error = 0.3064516129032258

Test Split = 0.3
Model coefficients: [0.01467657613765766, -0.03480718486738166, 0.08158880595023926]
Intercept = -0.000470198769377503
Predicted Statuses: [1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
Classification Error = 0.25806451612903225

Test Split = 0.2
Model coefficients: [0.014492882515963494, -0.005864287719611618, 0.07972172670130782]
Intercept = -1.8391308922488454
Predicted Statuses: [1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

Least classification error (0.258) obtained for Test Split = 0.3